faculty/clients/experiment.py

```diff
@@ -245,6 +334,98 @@ def make_restore_runs_response(self, data):
         return RestoreExperimentRunsResponse(**data)
 
 
+class SingleFilterValueField(fields.Field):
+    """
+    Field that serialises/deserialises a run filter.
+    """
+
+    def _is_valid_uuid(self, value, obj):
+        return isinstance(value, uuid.UUID) and (
+            obj.by in {SingleFilterBy.PROJECT_ID, SingleFilterBy.RUN_ID}
+        )
+
+    def _is_valid_experiment_id(self, value, obj):
+        return (
+            isinstance(value, int) and obj.by == SingleFilterBy.EXPERIMENT_ID
+        )
+
+    def _is_directly_stringifiable(self, value, obj):
+        return (
+            self._is_valid_uuid(value, obj)
+            or self._is_valid_experiment_id(value, obj)
+            or obj.by
+            in {
+                SingleFilterBy.TAG,
+                SingleFilterBy.PARAM,
+                SingleFilterBy.METRIC,
+            }
+        )
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return value
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if self._is_directly_stringifiable(value, obj):
+            return str(value)
+        elif obj.by == SingleFilterBy.DELETED_AT:
+            return marshmallow_utils.from_iso_datetime(str(value)).isoformat()
+        else:
+            raise RunQueryFilterValidation(
+                "Validation error serialising run query filter", value
+            )
+
+
+class FilterField(fields.Field):
+    """
+    Field that serialises/deserialises a run filter.
+    """
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        if value is None:
+            return None
+        elif isinstance(value, SingleFilter):
+            return SingleFilterSchema().load(value)
+        else:
+            return CompoundFilterSchema().load(value)
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        print(type(obj))
+        if value is None and isinstance(obj, QueryRuns):
```

Comment:
Why do we need to check the type of the obj here?

Reply:
It's because of the recursive nature of this serialisation:
- QueryRuns is allowed to have None as a filter
- However, compound filters are not allowed to have None filters in their conditions